src/ir/typescript_types.py

```diff
@@ -402,14 +424,375 @@ def gen_number_literal(self):
+            X: string. We remove X from all possible substitutions for other T1 types and then
+            delete the pair with key string from our possible_subs dict.
+
+        (4) We sort the dict again, it now looks like this:
+                sorted_type_subs = [(number, {Y})]
+
+        (5) Assign the substitution Y: number and repeat the rest of step (3)
+
+        (6) The possible_subs dict is now empty, so we return our substitution dictionary
+                return {X: string, Y: number}
+        """
+        type_var_map = {}
+
+        # Continue trying to find type variable susbstitutions until
+        # all T1 types are substituted in T2.
+        while possible_subs:
+            # Sort the possible_subs dict, in order to first find a substitution for the T1
+            # type with the fewest compatible T2 type variables.
+            sorted_type_subs = sorted(list(possible_subs.items()), key=lambda x: len(x[1]), reverse=False)
+
+            # Get the first (T1 type, T2 type variable) pair (sorted_type_subs is a tuples list)
+            type_to_substitute, compatible_tvars = sorted_type_subs[0]
+
+            # If there aren't any compatible_tvars, then that means that there is no possible
+            # order of substitutions that ensures all types in T1 are substituted in T2.
+            # Hence, we return a False flag to indicate that the type unification is not possible.
+            # Note: at this point this happens if at previous iterations of the while loop
+            # we substituted all the type variables that are compatible with this specific type_to_substitute.
+            if not compatible_tvars:
+                return ({}, False)
+
+            # Get any of the compatible type variables and substitute it with the T1 type
+            chosen_tvar = compatible_tvars.pop()
+            type_var_map[chosen_tvar] = type_to_substitute
+
+            # Remove the substituted type variable from the possible substitutions
+            # of all other T1 types.
+            for k in list(possible_subs.keys()):
+                if chosen_tvar in possible_subs[k]:
+                    possible_subs[k].remove(chosen_tvar)
+
+            # Delete the possible substitutions of the T1 type we just substituted in T2
+            del possible_subs[type_to_substitute]
+
+        # Return the substitutions we found and a flag confirming that there is a possible
+        # order of substitutions that gurantees type unification.
+        return (type_var_map, True)
+
+    def get_name(self):
+        return self.name
+
+    def __str__(self):
+        return f'UnionType(TS)({" | ".join([str(t) for t in self.types])})'
+
+    def __eq__(self, other):
+        return (self.__class__ == other.__class__ and
+                self.name == other.name and
+                set(self.types) == set(other.types))
+
+    def __hash__(self):
+        return hash(str(self.name) + str(self.types))
+
+
+class UnionTypeFactory:
```

Comment:
`UnionTypeFactory` => `UnionTypeFactory(object)`.

Reply:
Done as suggested by PEP8. The reason it is recommended is compatibility between Python 3 and 2. [Detailed explanation](https://stackoverflow.com/questions/4015417/why-do-python-classes-inherit-object).